kong/cluster_events/init.lua

```diff
@@ -326,80 +327,45 @@ if ngx_debug then
 end
 
 
-local function get_lock(self)
-  -- check if a poll is not currently running, to ensure we don't start
-  -- another poll while a worker is still stuck in its own polling (in
-  -- case it is being slow)
-  -- we still add an exptime to this lock in case something goes horribly
-  -- wrong, to ensure other workers can poll new events
-  -- a poll cannot take more than max(poll_interval * 5, 10) -- 10s min
-  local ok, err = self.shm:safe_add(POLL_RUNNING_LOCK_KEY, true,
-                                    max(self.poll_interval * 5, 10))
-  if not ok then
-    if err ~= "exists" then
-      log(ERR, "failed to acquire poll_running lock: ", err)
-    end
-    -- else
-    --   log(DEBUG, "failed to acquire poll_running lock: ",
-    --              "a worker still holds the lock")
-
-    return false
-  end
-
-  if self.poll_interval > 0.001 then
-    -- check if interval of `poll_interval` has elapsed already, to ensure
-    -- we do not run the poll when a previous poll was quickly executed, but
-    -- another worker got the timer trigger a bit too late.
-    ok, err = self.shm:safe_add(POLL_INTERVAL_LOCK_KEY, true,
-                                self.poll_interval - 0.001)
-    if not ok then
-      if err ~= "exists" then
-        log(ERR, "failed to acquire poll_interval lock: ", err)
-      end
-      -- else
-      --   log(DEBUG, "failed to acquire poll_interval lock: ",
-      --              "not enough time elapsed since last poll")
-
-      self.shm:delete(POLL_RUNNING_LOCK_KEY)
-
-      return false
-    end
-  end
-
-  return true
-end
-
-
 poll_handler = function(premature, self)
   if premature or not self.polling then
     -- set self.polling to false to stop a polling loop
     return
   end
 
-  if not get_lock(self) then
-    local ok, err = timer_at(self.poll_interval, poll_handler, self)
-    if not ok then
-      log(CRIT, "failed to start recurring polling timer: ", err)
+  -- check if a poll is not currently running, to ensure we don't start
+  -- another poll while a worker is still stuck in its own polling (in
+  -- case it is being slow)
+  -- we still add an exptime to this lock in case something goes horribly
+  -- wrong, to ensure other workers can poll new events
+  -- a poll cannot take more than max(poll_interval * 5, 10) -- 10s min
+  local ok, err = concurrency.with_worker_mutex({
+    name = POLL_RUNNING_LOCK_KEY,
+    timeout = 0,
+    exptime = max(self.poll_interval * 5, 10),
+  }, function()
+    if self.poll_interval > 0.001 then
+      -- check if interval of `poll_interval` has elapsed already, to ensure
+      -- we do not run the poll when a previous poll was quickly executed, but
+      -- another worker got the timer trigger a bit too late.
+      return concurrency.with_worker_mutex({
+        name = POLL_INTERVAL_LOCK_KEY,
+        timeout = 0,
+        exptime = self.poll_interval - 0.001,
+      }, function()
+        return poll(self)
+      end)
```

Comment:
One thing in `concurrency` is that it makes difficult to not use anonymous functions. This could be fixed by changing:
```lua
function concurrency.with_worker_mutex(opts, fn)
```
to
```lua
function concurrency.with_worker_mutex(opts, fn, ...)
```

and

```lua
function concurrency.with_coroutine_mutex(opts, fn)
```
to
```lua
function concurrency.with_coroutine_mutex(opts, fn, ...)
```

Worth to consider that, but I am also fine if done later, if it feels scope creep.

Reply:
at the moment `with_worker_mutex` passes exactly one argument (`elapsed`) to `fn`: https://github.com/Kong/kong/blob/4adb6772a13e036f770ef6110c665e663db8cade/kong/concurrency.lua#L59

If we change it as you say and still keep `elapsed` as one of the arguments to `fn` (the first?) then we should always remember that `fn` won't only receive the args that are passed as varargs to `with_worker_mutex`, which would also be inconsistent to how things work in `with_coroutine_mutex`. What do you think @bungle ?

Reply:
@samugi, should we always pass the elapsed? Even with coroutine mutex? Is it possible?

Reply:
> Semaphore api does not return it, so it needs some manual counting in concurrency to do that.

exactly @bungle . I would propose keeping the scope of this PR small otherwise it will be delayed because I won't have time to work on this soon, we can address your proposed improvements in a separate PR, how does that sound?